panic when loading non-exist geo file tag with the new geo file decoder

## What version of V2Ray are you using?

7e87670 or later

## What problems have you encountered?

panic when loading non-exist geo file tag with the new geo file decoder

## What's your expectation?

expected behavior on versions before 7e87670

```javascript
main: failed to read config files: [**Hide**\config.json] > infra/conf: invalid field rule > infra/conf: failed to parse domain rule: geosite:nonexist > infra/conf: failed to load geosite: nonexist > infra/conf: list not found in geosite.dat: nonexist
```

## Please attach your configuration here

```javascript
{
    "inbounds": [
        {
            "listen": "127.0.0.1",
            "port": 1080,
            "protocol": "socks",
            "settings": {}
        }
    ],
    "log": {
        "loglevel": "debug"
    },
    "outbounds": [
        {
            "protocol": "freedom",
            "settings": {},
            "tag": "direct"
        }
    ],
    "routing": {
        "rules": [
            {
                "domain": [
                    "geosite:nonexist"
                ],
                "outboundTag": "direct",
                "type": "field"
            }
        ]
    }
}

```

## Please attach error logs here

```javascript
[Warning] common/geodata: failed to decode geodata file: **Hide**\geosite.dat. Fallback to the original ReadFile method.
panic: runtime error: invalid memory address or nil pointer dereference
[signal 0xc0000005 code=0x0 addr=0x38 pc=0x1755314]

goroutine 1 [running]:
github.com/v2fly/v2ray-core/v4/common/geodata.LoadSite(0x19a4c8c, 0xb, 0xc0004eb998, 0x8, 0x0, 0x0, 0xc0004f49f0, 0x1, 0x1)
        github.com/v2fly/v2ray-core/v4/common/geodata/load.go:27 +0x1f4
github.com/v2fly/v2ray-core/v4/infra/conf.loadGeositeWithAttr(0x19a4c8c, 0xb, 0xc0004eb998, 0x8, 0xc0003c20b0, 0xc0004df798, 0x10196f4, 0xc000290a50, 0xc0004df7b0)
        github.com/v2fly/v2ray-core/v4/infra/conf/geodata.go:31 +0x330
github.com/v2fly/v2ray-core/v4/infra/conf.loadGeosite(...)
        github.com/v2fly/v2ray-core/v4/infra/conf/geodata.go:16
github.com/v2fly/v2ray-core/v4/infra/conf.parseDomainRule(0xc0004eb990, 0x10, 0xb0, 0x17e5040, 0xc0000fa640, 0x0, 0x0)
        github.com/v2fly/v2ray-core/v4/infra/conf/router.go:225 +0x1e5
github.com/v2fly/v2ray-core/v4/infra/conf.parseFieldRule(0xc0003c20b0, 0xae, 0xb0, 0x1, 0xc000096d01, 0x0)
        github.com/v2fly/v2ray-core/v4/infra/conf/router.go:441 +0xc05
github.com/v2fly/v2ray-core/v4/infra/conf.ParseRule(0xc0003c20b0, 0xae, 0xb0, 0xc0004dfa20, 0xecf125, 0xc0000ba980)
        github.com/v2fly/v2ray-core/v4/infra/conf/router.go:519 +0x1f2
github.com/v2fly/v2ray-core/v4/infra/conf.(*RouterConfig).Build(0xc0000d21e0, 0xc0000ba988, 0x3, 0xc00009d5d8)
        github.com/v2fly/v2ray-core/v4/infra/conf/router.go:98 +0x191
github.com/v2fly/v2ray-core/v4/infra/conf.(*Config).Build(0xc0003c4000, 0xc0000d0990, 0xc0003c40e0, 0x0)
        github.com/v2fly/v2ray-core/v4/infra/conf/v2ray.go:536 +0x1e11
github.com/v2fly/v2ray-core/v4/main/jsonem.init.0.func1(0x182c760, 0xc00009c1e0, 0xc0000872e3, 0x4, 0xc00003bc30)
        github.com/v2fly/v2ray-core/v4/main/jsonem/jsonem.go:36 +0xaf
github.com/v2fly/v2ray-core/v4.LoadConfig(0x199e988, 0x4, 0xc0000872c0, 0x27, 0x182c760, 0xc00009c1e0, 0x1ad7580, 0xc000006018, 0xc0004dfe88)
        github.com/v2fly/v2ray-core/v4/config.go:67 +0x26c
main.startV2Ray(0xc000060120, 0xc0000541f0, 0x0, 0x0)
        github.com/v2fly/v2ray-core/v4/main/main.go:110 +0x135
main.main()
        github.com/v2fly/v2ray-core/v4/main/main.go:139 +0xbf
```

--- a/common/geodata/cache.go
+++ b/common/geodata/cache.go
@@ -32,13 +32,13 @@ func (g GeoIPCache) Set(key string, value *router.GeoIP) {
 }
 
 func (g GeoIPCache) Unmarshal(filename, code string) (*router.GeoIP, error) {
-	filename = platform.GetAssetLocation(filename)
-	idx := strings.ToUpper(filename + "|" + code)
+	asset := platform.GetAssetLocation(filename)
+	idx := strings.ToUpper(asset + "|" + code)
 	if g.Has(idx) {
 		return g.Get(idx), nil
 	}
 
-	geoipBytes, err := Decode(filename, code)
+	geoipBytes, err := Decode(asset, code)
 	switch err {
 	case nil:
 		var geoip router.GeoIP
@@ -48,10 +48,13 @@ func (g GeoIPCache) Unmarshal(filename, code string) (*router.GeoIP, error) {
 		g.Set(idx, &geoip)
 		return &geoip, nil
 
+	case errCodeNotFound:
+		return nil, newError(code, " not found in ", filename)
+
 	case errFailedToReadBytes, errFailedToReadExpectedLenBytes,
 		errInvalidGeodataFile, errInvalidGeodataVarintLength:
 		newError("failed to decode geodata file: ", filename, ". Fallback to the original ReadFile method.").AtWarning().WriteToLog()
-		geoipBytes, err = ioutil.ReadFile(filename)
+		geoipBytes, err = ioutil.ReadFile(asset)
 		if err != nil {
 			return nil, err
 		}
@@ -72,7 +75,7 @@ func (g GeoIPCache) Unmarshal(filename, code string) (*router.GeoIP, error) {
 		return nil, err
 	}
 
-	return nil, nil
+	return nil, newError(code, " not found in ", filename)
 }
 
 type GeoSiteCache map[string]*router.GeoSite
@@ -96,13 +99,13 @@ func (g GeoSiteCache) Set(key string, value *router.GeoSite) {
 }
 
 func (g GeoSiteCache) Unmarshal(filename, code string) (*router.GeoSite, error) {
-	filename = platform.GetAssetLocation(filename)
-	idx := strings.ToUpper(filename + "|" + code)
+	asset := platform.GetAssetLocation(filename)
+	idx := strings.ToUpper(asset + "|" + code)
 	if g.Has(idx) {
 		return g.Get(idx), nil
 	}
 
-	geositeBytes, err := Decode(filename, code)
+	geositeBytes, err := Decode(asset, code)
 	switch err {
 	case nil:
 		var geosite router.GeoSite
@@ -112,10 +115,13 @@ func (g GeoSiteCache) Unmarshal(filename, code string) (*router.GeoSite, error) 
 		g.Set(idx, &geosite)
 		return &geosite, nil
 
+	case errCodeNotFound:
+		return nil, newError(code, " not found in ", filename)
+
 	case errFailedToReadBytes, errFailedToReadExpectedLenBytes,
 		errInvalidGeodataFile, errInvalidGeodataVarintLength:
 		newError("failed to decode geodata file: ", filename, ". Fallback to the original ReadFile method.").AtWarning().WriteToLog()
-		geositeBytes, err = ioutil.ReadFile(filename)
+		geositeBytes, err = ioutil.ReadFile(asset)
 		if err != nil {
 			return nil, err
 		}
@@ -136,5 +142,5 @@ func (g GeoSiteCache) Unmarshal(filename, code string) (*router.GeoSite, error) 
 		return nil, err
 	}
 
-	return nil, nil
+	return nil, newError(code, " not found in ", filename)
 }

--- a/common/geodata/decode.go
+++ b/common/geodata/decode.go
@@ -10,6 +10,7 @@
 package geodata
 
 import (
+	"io"
 	"os"
 	"runtime"
 	"strings"
@@ -26,6 +27,7 @@ var (
 	errFailedToReadExpectedLenBytes = errors.New("failed to read expected length of bytes")
 	errInvalidGeodataFile           = errors.New("invalid geodata file")
 	errInvalidGeodataVarintLength   = errors.New("invalid geodata varint length")
+	errCodeNotFound                 = errors.New("code not found")
 )
 
 func emitBytes(f *os.File, code string) ([]byte, error) {
@@ -41,6 +43,9 @@ Loop:
 	for {
 		container := make([]byte, advancedN)
 		bytesRead, err := f.Read(container)
+		if err == io.EOF {
+			return nil, errCodeNotFound
+		}
 		if err != nil {
 			return nil, errFailedToReadBytes
 		}

